sympy/core/containers.py

```diff
@@ -194,8 +195,9 @@ class Dict(Basic):
     """
 
     def __new__(cls, *args):
-        if len(args) == 1 and ((args[0].__class__ is dict) or
-                             (args[0].__class__ is Dict)):
+        if len(args) == 1 and (isinstance(args[0], dict) or
+            isinstance(args[0], Dict) or
+            isinstance(args[0], defaultdict)):
```

Comment:
> Is that condition necessary? An instance of defaultdict is also an
instance of dict.
>
> —

defaultdict is inherited from dict. But, `__class__` returns the type which
is `defaultdict`. So, it won't work.